File: report10/cgi-counter.rb
#!/usr/local3/coins/linux/bin/ruby
# -*- coding: utf-8 -*-
require "cgi"

DATA_FILENAME = "file-counter-value.data"
$comname = $0

def main()
  @cgi = CGI.new()
  print_header()
  counter =  handle_counter()
  print_content(counter)
  exit( 0 )
end

def print_header()
  printf("Content-Type: text/html\n")
  printf("\n")
end

def handle_counter()
  counter = counter_load()
  op = @cgi["op"]
  val = @cgi["val"].to_i()

  case(op)
  when "inc"
    counter += 1
    counter_save(counter)
  when "set"
    counter = val
    counter_save(counter)
  else
  end

  return counter
end

def print_content(counter)
  printf("<HTML><HEAD></HEAD><BODY>\n")
  $stdout.flush()
  printf("%d\n", counter)
  printf("</BODY></HTML>\n")
end

def counter_save(counter)
  data = [counter].pack("i") # signed int
  IO.binwrite(DATA_FILENAME, data)
end

def counter_load()
  counter = 0
  begin
    data = IO.read(DATA_FILENAME, 4)
    counter = data.unpack("i")[0]
  rescue Errno::ENOENT
    # nop
  end
    return counter
end

main()

File: report10/cgi-counter.py
#!/usr/local3/coins/linux/bin/python3
# -*- coding: utf-8 -*-

import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)

import cgi
import bleach
import sys

DATA_FILENAME = "file-counter-value.data"

def main():
    print_header()
    counter = handle_counter()
    print_content(counter)
    exit(0)

def print_header():
    print("Content-Type: text/html")
    print("")

def handle_counter():
    counter = counter_load()
    qh = cgi.FieldStorage(keep_blank_values=True)
    op = qh.getfirst("op","")
    if op == "inc":
        counter += 1
        counter_save(counter)
    elif op == "set":
        val = int(qh.getfirst("val",""))
        counter = val
        counter_save(counter)
    return counter
   
def print_content(counter):
    print("<HTML><HEAD></HEAD><BODY><PRE>",flush=True)
    print(counter)
    print("</PRE></BODY></HTML>\n")

def counter_load():
    counter = 0
    try:
        with open(DATA_FILENAME, "rb") as f:
            counter_bytes = f.read(4)
            counter = int.from_bytes(counter_bytes, byteorder=sys.byteorder)
    except FileNotFoundError:
        pass
    return counter

def counter_save(counter):
    counter_bytes = counter.to_bytes(4, byteorder=sys.byteorder)
    with open(DATA_FILENAME, "wb") as f:
        f.write(counter_bytes)
    return

main()
